client/tests/simulator_tests.rs

```diff
@@ -83,6 +85,66 @@ fn test_get_capability_manufacturer_id() {
     assert_eq!(received, expected);
 }
 
+#[test]
+fn test_get_random_max_size() {
+    let (_sim_lifeline, mut tpm) = get_started_tpm();
+
+    let command = GetRandomCmd {
+        bytes_requested: 4096,
+    };
+
+    // We allow panic in test cases.
+    let resp = run_command(&command, &mut tpm).expect("Failed running command.");
+
+    // Lets pull out the actual data as a slice for convenience
+    let random_slice = &resp.random_bytes.buffer[..resp.random_bytes.size as usize];
+
+    // We should have received less than what the client implementation supports.
+    assert!(random_slice.len() <= TpmtHa::UNION_SIZE);
+
+    // We should have received minimum size of SHA256 (minimum mandatory digest size).
+    assert!(random_slice.len() >= TPM2_SHA256_DIGEST_SIZE as usize);
+
+    // Print for test inspection.
+    trace!("Got random sequence: {:?}", random_slice);
+
+    // Duplicate value trap, to catch any same value sequences.
+    {
+        let mut same_twice: u16 = 0;
+        for i in 1..random_slice.len() as usize {
```

Comment:
Optional: Could be done with iterators; I find this more readable though not all may:

```rs
let same_twice_occurrences = random_slice.iter()
    .zip(random_slice.iter().skip(1))
    .filter(|(prev, next)| prev == next)
    .count();
```

---

Ideally Rust would provide a window function, which is what you really want here, but I don't know that I want to pull in itertools for a single function. Seems likely to someday be in the std so I guess we can just build the window manually for now using `zip`.

Reply:
Ah - another great Rust construct 🙂 I am still thinking too much in the "C" programming strategies. But here the Rust compiler will optimize (i hope) by it self. I have adopted this strategy - thanks again for treat tips 🙂 👍

Reply:
It should! Iterators are one of those types that attempt to inline and be written as-if you had written them by hand.

In general, you should feel empowered to use the helper functions of `Option`, `Result`, and `Iterator` as-if you were writing their inner code out by hand; this is very different from what we usually think in C and C++.

My experience has been the compiler is pretty good about this stuff. Most of the time producing something as if you had written the trickier for-loop out yourself. So I think we should trust the compiler for these constructs until we have a situation where we need to be *very certain* about the efficiency of something (and tests I'm not too worried about efficiency, so I'd opt for whatever is most readable here regardless of the cost, personally).

Anyways, this is one of the things that Rust means when it promises these "zero-cost abstractions".

Reply:
Thanks for the great clarification 👍 🙂